Added notes on OnRemove event

FriedrichWeinmann opened the pull request.

zjalexander commented:
### :x: Validation status: errors


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Core/Remove-Module.md](https://github.com/FriedrichWeinmann/PowerShell-Docs/blob/patch-1/reference/6/Microsoft.PowerShell.Core/Remove-Module.md) | :x:Error | | [Details](#user-content-reference%2F6%2FMicrosoft.PowerShell.Core%2FRemove-Module.md)
[]() | :x:Error | | [Details](#user-content-)

<a name="reference%2F6%2FMicrosoft.PowerShell.Core%2FRemove-Module.md"></a>
### [reference/6/Microsoft.PowerShell.Core/Remove-Module.md](https://github.com/FriedrichWeinmann/PowerShell-Docs/blob/patch-1/reference/6/Microsoft.PowerShell.Core/Remove-Module.md)
 - **[Error]** Unable to load file: 6/Microsoft.PowerShell.Core/Remove-Module.md via processor: AzureCliDocumentProcessor.

<a name=""></a>
### []()
 - **[Error]** Markdown.MAML.Transformer.HelpSchemaException: :198:(1218) &#39;```powershell
$PSF_OnRemoveScript = {
	Get-PSFRu...&#39;
 Expect Heading
   at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.GetInnerContexts(DocumentBuildParameters parameters, IEnumerable`1 processors, TemplateProcessor templateProcessor, IHostServiceCreator creator)
   at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.BuildCore(DocumentBuildParameters parameters)
   at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.BuildCore(DocumentBuildParameters parameter, IMarkdownServiceProvider markdownServiceProvider, BuildInfo currentBuildInfo, BuildInfo lastBuildInfo)
   at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.Build(IList`1 parameters, String outputDirectory)
   at Microsoft.DocAsCode.SubCommands.DocumentBuilderWrapper.BuildDocument(BuildJsonConfig config, TemplateManager templateManager, String baseDirectory, String outputDirectory, String pluginDirectory, String templateDirectory)
   at Microsoft.DocAsCode.SubCommands.DocumentBuilderWrapper.BuildDocument()

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C13%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201803131011393644-2174%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=sCSysn0usX9XikzShlGP%2B7lh1IJkx%2FGYo4RCxXR7Chk%3D&st=2018-03-13T10%3A32%3A45Z&se=2018-04-13T10%3A37%3A45Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

zjalexander commented:
### :x: Validation status: errors


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Core/Remove-Module.md](https://github.com/FriedrichWeinmann/PowerShell-Docs/blob/patch-1/reference/6/Microsoft.PowerShell.Core/Remove-Module.md) | :x:Error | | [Details](#user-content-reference%2F6%2FMicrosoft.PowerShell.Core%2FRemove-Module.md)
[]() | :x:Error | | [Details](#user-content-)

<a name="reference%2F6%2FMicrosoft.PowerShell.Core%2FRemove-Module.md"></a>
### [reference/6/Microsoft.PowerShell.Core/Remove-Module.md](https://github.com/FriedrichWeinmann/PowerShell-Docs/blob/patch-1/reference/6/Microsoft.PowerShell.Core/Remove-Module.md)
 - **[Error]** Unable to load file: 6/Microsoft.PowerShell.Core/Remove-Module.md via processor: AzureCliDocumentProcessor.

<a name=""></a>
### []()
 - **[Error]** Markdown.MAML.Transformer.HelpSchemaException: :198:(1218) &#39;```
$PSF_OnRemoveScript = {
	Get-PSFRunspace | S...&#39;
 Expect Heading
   at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.GetInnerContexts(DocumentBuildParameters parameters, IEnumerable`1 processors, TemplateProcessor templateProcessor, IHostServiceCreator creator)
   at Microsoft.DocAsCode.Build.Engine.SingleDocumentBuilder.BuildCore(DocumentBuildParameters parameters)
   at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.BuildCore(DocumentBuildParameters parameter, IMarkdownServiceProvider markdownServiceProvider, BuildInfo currentBuildInfo, BuildInfo lastBuildInfo)
   at Microsoft.DocAsCode.Build.Engine.DocumentBuilder.Build(IList`1 parameters, String outputDirectory)
   at Microsoft.DocAsCode.SubCommands.DocumentBuilderWrapper.BuildDocument(BuildJsonConfig config, TemplateManager templateManager, String baseDirectory, String outputDirectory, String pluginDirectory, String templateDirectory)
   at Microsoft.DocAsCode.SubCommands.DocumentBuilderWrapper.BuildDocument()

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C13%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201803131258135146-2174%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=3cvVwU8rWrX8TTNEBFs7pwZoRCuRWvqT49j9VcZsXRY%3D&st=2018-03-13T13%3A16%3A49Z&se=2018-04-13T13%3A21%3A49Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

FriedrichWeinmann commented:
Can anybody explain to me what is happening here that it is failing?
As far as I can see I PRed legal markup text, and I do not have access to whatever it is you are using to compile it.

sdwheeler commented:
@FriedrichWeinmann This is a PlatyPS schema violation. According to the schema you cannot have code blocks in the Notes section.

FriedrichWeinmann commented:
@sdwheeler thanks for clarifying this. While I disagree with the design decision behind that limitation, it is none-the-less a barrier to live with. Updated it and removed the block notation.

sdwheeler commented:
@FriedrichWeinmann I agree. I don't like the design either. We need an update to PlatyPS.

zjalexander commented:
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Core/Remove-Module.md](https://github.com/FriedrichWeinmann/PowerShell-Docs/blob/patch-1/reference/6/Microsoft.PowerShell.Core/Remove-Module.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C13%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201803131718486984-2174%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=QGiCDs2jmoJ9hbjzUDShp9qAb%2Ba0YhK9Ym6FSbSQmxQ%3D&st=2018-03-13T17%3A42%3A25Z&se=2018-04-13T17%3A47%3A25Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

sdwheeler commented:
@SteveL-MSFT Can you review this change?

SteveL-MSFT commented on reference/6/Microsoft.PowerShell.Core/Remove-Module.md:
```diff
@@ -205,6 +205,20 @@ You can pipe module names and module objects to **Remove-Module**.
 This cmdlet does not generate any output.
 
 ## NOTES
+When removing a module, there is an event on the module that will execute. This allows a module to react to being removed and maybe free up resources. Example:
```
Please use [semantic line breaks](http://sembr.org/) which basically means putting each sentence on a separate line.  Rendering will put it back together in paragraphs.

SteveL-MSFT replied:
I think the second sentence should be:

```none
This event allows a module to react to being removed and perform some cleanup such as freeing up resources.
```

SteveL-MSFT commented on reference/6/Microsoft.PowerShell.Core/Remove-Module.md:
```diff
@@ -205,6 +205,20 @@ You can pipe module names and module objects to **Remove-Module**.
 ## NOTES
+When removing a module, there is an event on the module that will execute. This allows a module to react to being removed and maybe free up resources. Example:
+
+$PSF_OnRemoveScript = {
```
I realize this example comes from your PSFramework module, but I think it can be a bit confusing to new users and would suggest making it more generic:

```powershell
$OnRemoveScript = {
  # perform cleanup
  $cachedSessions | Remove-PSSession
}

$ExecutionContext.SessionState.Module.OnRemove += $OnRemoveScript
```

FriedrichWeinmann replied:
Absolutely, thanks for pointing this out - a bit of mental laziness crept up on me there. Updated, I hope the linebreaks are ok in that scriptblock, given the limitations of platyPS.

SteveL-MSFT commented on reference/6/Microsoft.PowerShell.Core/Remove-Module.md:
```diff
@@ -205,6 +205,20 @@ You can pipe module names and module objects to **Remove-Module**.
+
+$ExecutionContext.SessionState.Module.OnRemove += $PSF_OnRemoveScript
+
+For full consistency, it might be also useful to react to the closing of the powershell process:
```
PowerShell should have correct casing

SteveL-MSFT commented on reference/6/Microsoft.PowerShell.Core/Remove-Module.md:
```diff
@@ -205,6 +205,20 @@ You can pipe module names and module objects to **Remove-Module**.
+
+For full consistency, it might be also useful to react to the closing of the powershell process:
+
+Register-EngineEvent -SourceIdentifier ([System.Management.Automation.PsEngineEvent]::Exiting) -Action $PSF_OnRemoveScript
```
Change $PSF_OnRemoveScript to just $OnRemoveScript

sdwheeler commented:
@FriedrichWeinmann can you incorporate Steve's feedback? Thanks

FriedrichWeinmann commented:
@sdwheeler : done

zjalexander commented:
### :x: Validation status: errors


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[]() | :x:Error | | [Details](#user-content-)

<a name=""></a>
### []()
 - **[Error]** filemap T:/20180316-012816-69196048/qnzk54d4.bj0 doesn&#39;t exist.
 - **Line 46**: **[Error]** Call W:\qnaf-s\.optemp\packages\opbuild.filemappreprocessor.1.46.2\tools\Microsoft.OpenPublishing.Build.Applications.FileMapPreProcessor.exe to preprocess T:/20180316-012816-69196048/0ln3h0cv.kiy,T:/20180316-012816-69196048/t4a5xm44.tzw,T:/20180316-012816-69196048/1mb03ra1.1ld,T:/20180316-012816-69196048/oktvfipk.2ov,T:/20180316-012816-69196048/dq2myrre.5t1,T:/20180316-012816-69196048/hcymkqjk.lrb,T:/20180316-012816-69196048/m31ccygl.ivl,T:/20180316-012816-69196048/geyuw5kj.go2,T:/20180316-012816-69196048/kzakm5zq.mzn,T:/20180316-012816-69196048/mke5rynv.p1v,T:/20180316-012816-69196048/2sctdxip.zya,T:/20180316-012816-69196048/n5rg1s2u.j4f,T:/20180316-012816-69196048/qaqmqawq.1ns,T:/20180316-012816-69196048/fmiuceo2.tcg,T:/20180316-012816-69196048/qnzk54d4.bj0,T:/20180316-012816-69196048/wns1kvx2.ajk,T:/20180316-012816-69196048/spv5nakw.w4d,T:/20180316-012816-69196048/mfxq5ryx.sgc,T:/20180316-012816-69196048/2foseawq.fnv failed
 - **Line 334**: **[Error]** Failed to preprocess filemaps
 - **[Warning]** Failed to restore filemap with depot name: MSDN.sharepoint-ps, locale: en-us, branch: master, details: The operation has timed out.

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C16%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201803160127550302-2174%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=lIb0Ycj2ocE9VLKGhikqWFdyBZgitJ3lu8db1vOBgIQ%3D&st=2018-03-16T01%3A26%3A25Z&se=2018-04-16T01%3A31%3A25Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.

FriedrichWeinmann commented:
Um ... I'm sure those messages are really meaningful to somebody, but ... if it's something I did that broke things, I'd appreciate a translation 😅

sdwheeler commented:
@FriedrichWeinmann no that was not you. I think we have a bad VM instance in our build pool. I will take care of it.

sdwheeler commented on reference/6/Microsoft.PowerShell.Core/Remove-Module.md:
```diff
@@ -205,6 +205,23 @@ You can pipe module names and module objects to **Remove-Module**.
+
+$OnRemoveScript = {
+
+  # perform cleanup
```
Need to escape the `#` character. Markdown treats it as an H1 header. Change it to `\#`. We would have this problem in a code block. But you know the problem there.

sdwheeler commented:
@FriedrichWeinmann One more edit. See my review notes. When you submit it should kick off a new build.

FriedrichWeinmann commented:
Done :)

zjalexander commented:
### :white_check_mark: Validation status: passed


File | Status | Preview URL | Details
---- | ------ | ----------- | -------
[reference/6/Microsoft.PowerShell.Core/Remove-Module.md](https://github.com/FriedrichWeinmann/PowerShell-Docs/blob/patch-1/reference/6/Microsoft.PowerShell.Core/Remove-Module.md) | :white_check_mark:Succeeded | |

For more details, please refer to the [build report](https://opbuildstorageprod.blob.core.windows.net/report/2018%5C3%5C16%5Cbab27324-4c59-a348-1473-9c6d587e4020%5CPullRequest%5C201803160829413003-2174%5Cworkflow_report.html?sv=2015-02-21&sr=b&sig=wPwIZlTNXl3B8KviVOuhX0tTrr4IoulmbUZJFoBzVSw%3D&st=2018-03-16T08%3A53%3A41Z&se=2018-04-16T08%3A58%3A41Z&sp=r).

**Note:** If you changed an existing file name or deleted a file, broken links in other files to the deleted or renamed file are listed only in the full build report.
